src/local-mode.ts

```diff
@@ -56,51 +102,159 @@ export class LocalModeManager {
+		);
+		await this.cleanEngineSnapshots();
+		return this.db.models('engineSnapshot').insert({
+			snapshot: JSON.stringify(record.snapshot),
+			timestamp,
+		});
+	}
+
+	// Read the latest stored snapshot from the database.
+	public async retrieveLatestSnapshot(): Promise<EngineSnapshotRecord | null> {
+		const r = await this.db
+			.models('engineSnapshot')
+			.select()
+			.orderBy('rowid', 'DESC')
+			.first();
+
+		if (!r) {
+			return null;
+		}
+		return new EngineSnapshotRecord(
+			EngineSnapshot.fromJSON(r.snapshot),
+			new Date(Date.parse(r.timestamp)),
+		);
+	}
+
+	private async removeLocalModeArtifacts(objects: EngineSnapshot) {
+		log.debug(`Going to delete the following objects: ${objects}`);
+
+		// Delete engine objects. We catch every deletion error, so that we can attempt other objects deletions.
+		await Bluebird.map(objects.containers, cId => {
+			return this.docker
+				.getContainer(cId)
+				.remove({ force: true })
+				.catch(e => log.error(`Unable to delete container ${cId}`, e));
+		});
+		await Bluebird.map(objects.images, iId => {
+			return this.docker
+				.getImage(iId)
+				.remove({ force: true })
+				.catch(e => log.error(`Unable to delete image ${iId}`, e));
+		});
+		await Bluebird.map(objects.networks, nId => {
+			return this.docker
+				.getNetwork(nId)
+				.remove()
+				.catch(e => log.error(`Unable to delete network ${nId}`, e));
+		});
+		await Bluebird.map(objects.volumes, vId => {
+			return this.docker
+				.getVolume(vId)
+				.remove()
+				.catch(e => log.error(`Unable to delete volume ${vId}`, e));
+		});
+
+		// Remove any local mode state added to the database.
+		await this.db
+			.models('app')
+			.del()
+			.where({ source: 'local' })
+			.catch(e =>
+				log.error('Cannot delete local app entries in the database', e),
+			);
+	}
+
+	// Handle local mode state change.
+	// Input parameter is a target (new) state.
+	public async handleLocalModeStateChange(local: boolean) {
+		try {
+			const currentRecord = await this.collectEngineSnapshot();
+			if (local) {
+				return await this.storeEngineSnapshot(currentRecord);
+			}
+
+			const previousRecord = await this.retrieveLatestSnapshot();
+			if (!previousRecord) {
+				log.info('Previous engine snapshot was not stored. Skipping cleanup.');
+				return;
+			}
+
+			log.debug(
+				`Leaving local mode and cleaning up objects since ${previousRecord.timestamp.toISOString()}`,
+			);
+			await this.removeLocalModeArtifacts(
+				currentRecord.snapshot.diff(previousRecord.snapshot),
+			);
+			await this.cleanEngineSnapshots();
+		} catch (e) {
+			log.error(
+				`Problems managing engine state on local mode switch. Local mode: ${local}.`,
+				e,
+			);
+		} finally {
+			log.debug('Handling of local mode switch is completed');
+		}
+	}
+
+	// Returns a promise to await local mode switch completion started previously.
+	public async switchCompletion() {
+		if (!this.switchInProgress) {
```

Comment:
Generally the pattern we use here is `if (this.switchInProgress == null)`. Two reasons for this, one it catches both `null` and `undefined` (as does your method), but the second is that it's very clear what we're *actually* checking for.

Reply:
Changed to `== null`.